zip-0321.rst

```diff
@@ -0,0 +1,178 @@
+
+The terms below are to be interpreted as follows:
+
+payment
+  A transfer of funds implemented by a shielded or transparent output of 
+  a single Zcash transaction.  
+
+
+Abstract
+========
+
+This ZIP proposes a standard format for payment request URIs. Wallets that 
+recognize this format enable users to construct payments simply by
+clicking links on webpages or scanning QR codes.
+
+
+Motivation
+==========
+
+In order for a robust transactional ecosystem to evolve for Zcash, it is
+necessary for vendors to be able to issue requests for payment. At present, the
+best option available is to manually specify a payment address, a payment
+amount, and potentially memo field content. Of these three components, existing
+wallets only provide functionality for reading payment addresses in a
+semi-automated fashion. It is then necessary for the user to manually enter
+payment amounts and any associated memo information, which is tedious and may
+be error-prone, particularly if a payment is intended for multiple recipients
+or the memo field information contains structured data that must be faithfully
+reproduced. 
+
+This ZIP seeks to eliminate these issues by proposing a standard format that
+wallet vendors may support so that human intervention is required only for
+approval, not creation, of such a payment transaction.
+
+In Bitcoin, two different standards exist to permit vendors to issue payment
+requests that are understood by wallets: BIP-0021 [#BIP0021]_ and BIP-0070
+[#BIP0070]_.  BIP-0021 provides a URI format that can be interpreted by a
+wallet to construct simple, single-recipient transactions; while BIP-0070 uses a
+protobuf-based protocol that permits payment requests to specify the creation
+of transactions of arbitrary complexity.
+
+The format proposed in this ZIP seeks a middle ground between these approaches:
+to provide a URI-based format which supports both the trivial use case and
+the slightly-more-complex situation where a payment may be intended for
+multiple recipients. 
+
+Requirements
+============
+
+The format must be a valid URI.
+
+The format must permit the representation of one or more (payment address, amount, 
+memo) tuples.
+
+Specification
+=============
+
+The following syntax specification uses ABNF [#RFC5234]_.
+
+URI Syntax:
+
+.. code-block:: EBNF
+
+  zcashurn        = "zcash:" ( zcashaddress [ "?" zcashparams ] / "?" zcashparams )
+  zcashaddress    = 1*base58
+  zcashparams     = zcashparam [ "&" zcashparams ]
+  zcashparam      = [ addrparam / amountparam / memoparam / messageparam / labelparam / reqparam / otherparam ]
+  NONZERO         = %x31-39
+  DIGIT           = %x30-39
+  paramindex      = "." NONZERO *DIGIT
+  addrparam       = "address" [ paramindex ] "=" zcashaddress
+  amountparam     = "amount"  [ paramindex ] "=" 1*DIGIT [ "." 1*DIGIT ]
+  labelparam      = "label"   [ paramindex ] "=" *qchar
+  memoparam       = "memo"    [ paramindex ] "=" *qchar
```

Comment:
It occurs to me here that maybe *qchar isn't the correct type for the memo parameter; we want to be able to encode 512 arbitrary bytes, and from my sample generation URI-encoded unicode (and anything else apart from ASCII) looks like the dog's breakfast. Would it make sense here to standardize on base58-encoded bytes (since we use base58 encoding elsewhere)? cc/ @daira

Reply:
We only use base58 for Sprout addresses. I think we should use base64.

Reply:
One potential issue with using base64 is that it specifies `=` for padding bytes, and would need to be hex-encoded. Base58 doesn't have this problem.

Reply:
The `=` padding is not actually needed; you can unambiguously infer the input length from the encoded length even without padding.

Reply:
Does this make it more complicated to specify/implement though? IIRC most base64 libraries automatically add padding, so implementers would have to be cautious of this.